# **01 - ETL & Data Cleaning**

## Objectives

- Load and inspect the raw Online Retail Transaction dataset.
- Assess data quality by identifying missing values, duplicate records, invalid quantities and unit prices.
- Clean and transform the dataset to prepare it for analysis.
- Create a `Revenue` feature to support further analysis.
- Save the cleaned dataset for exploratory data analysis and visualisation.

## Inputs

- Raw Online Retail Transaction dataset stored in the `data` folder.

## Outputs

- Cleaned Online Retail Transaction dataset.
- Cleaned data ready for exploratory data analysis and visualisation.

## Additional Comments

- Data-cleaning decisions will be based on the findings from the initial data inspection and documented throughout the notebook.

---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [ ]:
import os

current_dir = os.getcwd()
current_dir

'/Users/sahraosman/Documents/vscode-projects/Online Retail Transaction Analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/sahraosman/Documents/vscode-projects/Online Retail Transaction Analysis'

# Section 1 -  Import libraries

Here I will be importing the libraries that I will be using in this notebook.

In [4]:
import numpy as np
import pandas as pd


---

# Section 2 - Load the dataset

This section will load the raw Online Retail Transaction dataset and perform an initial inspection to understand its structure.

In [9]:
df = pd.read_csv("data/Online_Retail.csv")
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [11]:
print("Number of rows in the dataset:", df.shape[0])
print("Number of columns in the dataset:", df.shape[1])

Number of rows in the dataset: 541909
Number of columns in the dataset: 8


---

# Section 3 - Initial Data Inspection

Here I'll perform an initial inspection of the dataset to understand its structure, identify any data quality issues, and determine the necessary cleaning steps.

In [12]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   541909 non-null  int64  
 7   Country      541909 non-null  str    
dtypes: float64(1), int64(2), str(5)
memory usage: 33.1 MB


Looking at the structure of the dataset, we can see that it contains 8 columns and 541,909 rows. The columns are as follows:
InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, and Country.

I can already see that there are some missing values in the Description column. I will need to investigate further to determine the extent of the missing values and decide on an appropriate strategy for handling them.

InvoiceDate is currently in string format, and I will need to convert it to a datetime format for analysis.

In [ ]:
df.describe(include="all")

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
count,541909,541909,540455,541909.000000,541909,541909.000000,541909.000000,541909
unique,25900,4070,4223,NaN,23260,NaN,NaN,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,NaN,2011-10-31 14:41:00,NaN,NaN,United Kingdom
freq,1114,2313,2369,NaN,1114,NaN,NaN,495478
mean,NaN,NaN,NaN,9.552250,NaN,4.611114,15287.518434,NaN
std,NaN,NaN,NaN,218.081158,NaN,96.759853,1484.746041,NaN
min,NaN,NaN,NaN,-80995.000000,NaN,-11062.060000,12346.000000,NaN
25%,NaN,NaN,NaN,1.000000,NaN,1.250000,14367.000000,NaN
50%,NaN,NaN,NaN,3.000000,NaN,2.080000,15287.000000,NaN
75%,NaN,NaN,NaN,10.000000,NaN,4.130000,16255.000000,NaN


The descriptive statistics provide an overview of both numerical and categorical variables. The dataset contains 25,900 unique invoices, 4,070 unique stock codes and transactions across 38 countries. The United Kingdom accounts for the majority of transaction records.

The Quantity and UnitPrice columns contain negative minimum values, with Quantity ranging from -80,995 to 80,995 and UnitPrice ranging from -11,062.06 to 38,970. These values require further investigation during data cleaning.

The Description count is lower than the total number of rows, confirming the presence of missing values. Further analysis will be performed to quantify and investigate these records.

---

# Conclusions & Next Steps